Embrace hooks

mu1664al-s opened the pull request:
Changed as much as possible to use functional components with react hooks. Upgraded flow to support react hooks and adapted the tests to work with the changes. All tests are currently passing without any problem. We should stick to one convention from now on. Every new component should be functional and using react hooks. Stick to constant functional statements using arrow functions. Such as "const test = (param: type) => { // do something}" as it's immutable and arrow functions are self-binding in React, meaning we wont need to bind the references of functions to specific components in order to avoid context issues in the doom tree.

Feel free to take a look at the changes before merging it in to master.

adriangoransson commented:
This is a massive change, and it's not feasible to go through line by line. We'll take it for a spin tomorrow!

* What tools have been used to refactor? And how much has been done manually?
* Is there anything in particular that should be tested?

mu1664al-s commented:
> This is a massive change, and it's not feasible to go through line by line. We'll take it for a spin tomorrow!
> 
>     * What tools have been used to refactor? And how much has been done manually?
> 
>     * Is there anything in particular that should be tested?

It was all done manually. As far as I know, there are no tools for this. There are no need for changes between components, as the components are equivalent. As for the components, it's just a matter of conventions and moving around a little code. The translation is pretty straight forward. I was careful to maintain the internal logic.

Minor changes where done to the tests. I removed a test that checked a state value, as functional components doesn't reveal their internal state. It was a bad test to start with, these tests should be testing the outcomes only. The internal logic can always change. For some of the other tests, I had to add a delay because React hooks are asynchronous for efficiency.

What you could do is to temporary change the port of this branch and run this one and master side by side to compare the results.

I also added a test environment using Travis CI. The web app will be automatically built and tested on GitHub, whenever someone pushes up some code.

adriangoransson commented on src/Screens/Admin/Category/Category.js:
```diff
@@ -12,37 +12,38 @@ type Props = {
+  getCategory
+}: Props) => {
+  
+  useEffect(() => {
```
Empty dependency lists and then using the dependencies trigger warnings. There are multiple warnings.

Please verify that it's safe to use and if not maybe use `useCallback()`?
https://reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies

mu1664al-s replied:
This part is done

adriangoransson commented:
`/admin/deadlines` is broken.

mu1664al-s commented:
> `/admin/deadlines` is broken.

Done
